editor/filesystem_dock.cpp

```diff
@@ -3256,6 +3379,51 @@ void FileSystemDock::_file_and_folders_fill_popup(PopupMenu *p_popup, const Vect
 			p_popup->add_icon_item(get_editor_theme_icon(SNAME("NonFavorite")), TTR("Remove from Favorites"), FILE_REMOVE_FAVORITE);
 		}
 
+		if (p_paths.size() > 1 || p_paths[0] != "res://") {
+			cached_valid_conversion_targets = _get_valid_conversions_for_file_paths(p_paths);
+
+			int relative_id = 0;
+			if (!cached_valid_conversion_targets.is_empty()) {
+				p_popup->add_separator();
+
+				// If we have more the one type we can convert into, collapse it into a submenu.
+				const int CONVERSION_SUBMENU_THRESHOLD = 1;
+
+				if (cached_valid_conversion_targets.size() > CONVERSION_SUBMENU_THRESHOLD) {
+					PopupMenu *new_menu = memnew(PopupMenu);
+					new_menu->set_name("Convert to");
+					new_menu->connect("id_pressed", callable_mp(this, &FileSystemDock::_generic_rmb_option_selected));
+
+					p_popup->add_child(new_menu);
+					p_popup->add_submenu_item(TTR("Convert to..."), "Convert to", FILE_NEW);
+
+					for (const String &E : cached_valid_conversion_targets) {
+						Ref<Texture2D> icon;
+						if (has_theme_icon(E, SNAME("EditorIcons"))) {
+							icon = get_editor_theme_icon(E);
+						} else {
+							icon = get_theme_icon(E, SNAME("Resource"));
+						}
+
+						new_menu->add_icon_item(icon, vformat(TTR("%s"), E), CONVERT_BASE_ID + relative_id);
+						relative_id++;
+					}
+				} else {
+					for (const String &E : cached_valid_conversion_targets) {
```

Comment:
You could unify this loop with the one above by using `p_popup` in place of `new_menu` (probably rename the variable then).